File: themazeii.js
//Objective is similar to 'The Maze' but this time we want to find the shortest
//distance from point A to B.

let maze = 
[[0, 0, 1, 0, 0],
 [0, 0, 0, 0, 0],
 [0, 0, 0, 1, 0],
 [1, 1, 0, 1, 1],
 [0, 0, 0, 0, 0]]


//O(m*n) solution where m is the number of rows and n is the number of columns.
//We find the shortest distance at each pixel all the way until we reach the end

let directions = [[-1,0],[1,0],[0,1],[0,-1]]
    
start.push(0)
let queue = [start]

let distance = new Array(maze.length)
for (let i = 0; i < maze.length; i++) {
    distance[i] = new Array(maze[i].length).fill(Infinity)
}

while (queue.length > 0) {
    let [x,y,currDis] = queue.shift() 
    
    for (let [dx,dy] of directions) {
        let i = x
        let j = y 
        let tempDis = currDis
        
        while (i >= 0 && i < maze.length && j >= 0 && j < maze[0].length && maze[i][j] !== 1) {
            i += dx
            j += dy
            tempDis++
        }
        
        i -= dx
        j -= dy
        tempDis--
        
        if (distance[i][j] > tempDis) {
            distance[i][j] = tempDis
            queue.push([i, j, tempDis])
        }
    }
}

return distance[destination[0]][destination[1]] == Infinity ? -1 : distance[destination[0]][destination[1]]

File: themazeii.java
//Java Solution

class Solution {
    public int shortestDistance(int[][] maze, int[] start, int[] destination) {
        int[][] distances = new int[maze.length][maze[0].length];
        for (int i = 0; i < distances.length; i++) {
            Arrays.fill(distances[i], Integer.MAX_VALUE);
        }
        
        int[][] directions = new int[][] {{1,0}, {-1,0}, {0,1}, {0,-1}};
        
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[] {start[0], start[1], 0});
        
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            
            for (int i = 0; i < directions.length; i++) {
                int j = curr[0];
                int k = curr[1];
                int tempDis = curr[2];
                
                while (j >= 0 && j < maze.length && k >= 0 && k < maze[0].length && maze[j][k] != 1) {
                    j += directions[i][0];
                    k += directions[i][1];
                    tempDis++;
                }
                
                j -= directions[i][0];
                k -= directions[i][1];
                tempDis--;
                
                if (distances[j][k] > tempDis) {
                    distances[j][k] = tempDis;
                    queue.add(new int[] {j, k, tempDis});
                }
            }
        }
        
        return distances[destination[0]][destination[1]] == Integer.MAX_VALUE ? -1 : distances[destination[0]][destination[1]];
    }
}